scripts/wrapper/withdraw.js

```diff
@@ -12,34 +13,68 @@ module.exports = function (web3, artifacts) {
       throw new Error("Argument error: --requestWithdraw cannot be used with any of --withdraw, --transferFundsToMaster")
     }
+
+    if (!argv.withdrawalFile && !argv.from) {
+      throw new Error("Argument error: one of --withdrawalFile, --from must be given")
+    } else if (argv.withdrawalFile && argv.from) {
+      throw new Error("Argument error: --from cannot be used with --withdrawalFile")
+    }
+
+    if (argv.from) {
+      if (!argv.tokens && !argv.tokenIds) {
+        throw new Error("Argument error: one of --tokens, --tokenIds must be given when using --from")
+      } else if (argv.tokens && argv.tokenIds) {
+        throw new Error("Argument error: only one of --tokens, --tokenIds is required when using --from")
+      }
+    } else {
+      if (argv.tokens || argv.tokenIds) {
+        throw new Error("Argument error: --tokens or --tokenIds can only be used with --from")
+      }
+    }
   }
 
-  const getMaxWithdrawableAmount = async function (argv, bracketAddress, tokenInfo, exchange, printOutput = false) {
+  const getMaxWithdrawableAmount = async function (
+    argv,
+    bracketAddress,
+    tokenData,
+    exchange,
+    currentBatchId,
+    printOutput = false
+  ) {
     const log = printOutput ? (...a) => console.log(...a) : () => {}
     let amount
-    const token = tokenInfo.instance
-    if (argv.requestWithdraw) amount = (await exchange.getBalance(bracketAddress, tokenInfo.address)).toString()
+    const token = tokenData.instance
+    if (argv.requestWithdraw) amount = (await exchange.getBalance(bracketAddress, tokenData.address)).toString()
     else if (argv.withdraw) {
-      const currentBatchId = Math.floor(Date.now() / (5 * 60 * 1000)) // definition of BatchID, it avoids making a web3 request for each withdrawal to get BatchID
-      const pendingWithdrawal = await exchange.getPendingWithdraw(bracketAddress, tokenInfo.address)
-      if (pendingWithdrawal[1].toNumber() == 0) {
-        log("Warning: no withdrawal was requested for address", bracketAddress, "and token", tokenInfo.symbol)
-        amount = "0"
-      }
-      if (amount != "0" && pendingWithdrawal[1].toNumber() >= currentBatchId) {
-        log("Warning: amount cannot be withdrawn from the exchange right now, withdrawing zero")
+      const pendingWithdrawal = await exchange.getPendingWithdraw(bracketAddress, tokenData.address)
+      amount = pendingWithdrawal[0].toString()
+      if (pendingWithdrawal[1].toNumber() >= currentBatchId) {
+        const batchIdsLeft = pendingWithdrawal[1].toNumber() - currentBatchId + 1
+        log(
+          "Warning: requested withdrawal of " +
+            amount +
+            " " +
+            tokenData.symbol +
+            " for bracket " +
+            bracketAddress +
+            " cannot be executed until " +
+            batchIdsLeft +
+            " " +
+            (batchIdsLeft == 1 ? "batch" : "batches") +
+            " from now, skipping"
+        )
```

Comment:
I have to say that this looks pretty horrendous. I think you might be able to get away with some cleaner string formatting here. Are you familiar with the backtick string formatting?

```
`Warning: requested withdrawal of ${amount} ${tokenData.symbol} ...etc
```

Should work here...

Reply:
Nice idea, it's definitely more readable.

Reply:
Thanks so very much! I really like the way it looks now.